### PR TITLE
Fix smartcrop rotation problem

### PR DESCRIPTION
**Pull request by roborourke**
Right now smartcropping is applied prior to image rotation normalisation. This means we can end up with a bad extract area in sharp.

This update performs the rotation normalisation before any other transformations and also simplifies the resizeBuffer method using async/await.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -39,184 +39,158 @@ module.exports.s3 = function(config, key, args, callback) {
...
-						return;
-					});
-				});
+// return a default compression value based on a logarithmic scale
+// defaultValue = 100, zoom = 2; = 65
+// defaultValue = 80, zoom = 2; = 50
+// defaultValue = 100, zoom = 1.5; = 86
+// defaultValue = 80, zoom = 1.5; = 68
+const applyZoomCompression = function( defaultValue, zoom ) {
+	return clamp( Math.round( defaultValue - ( (Math.log(zoom) / Math.log(defaultValue / zoom)) * (defaultValue * zoom) ) ), Math.round(defaultValue / zoom), defaultValue );
```
Comment (rmccue) [inline index.js]: Spacin' needs to be consistent here

Also, maybe wrap the lines?

File: index.js [hunk 2]
```diff
@@ -39,184 +39,158 @@ module.exports.s3 = function(config, key, args, callback) {
...
+		image.rotate();
+
+		// convert gifs to pngs unless animated
+		if (
+			args.key &&
+			path.extname(args.key).toLowerCase() === '.gif'
+		) {
+			if (isAnimated(buffer)) {
+				return callback(new Error('fallback-to-original'));
+			} else {
```
Comment (rmccue) [inline index.js]: Could return early; but, error carried through from previous

File: index.js [hunk 3]
```diff
@@ -39,184 +39,158 @@ module.exports.s3 = function(config, key, args, callback) {
...
+			var cropValues =
+				typeof args.crop === 'string'
+					? args.crop.split(',')
+					: args.crop;
+
+			// convert percentages to px values
+			cropValues = cropValues.map(function(value, index) {
+				if (value.indexOf('px') > -1) {
+					return Number(value.substr(0, value.length - 2));
+				} else {
```
Comment (rmccue) [inline index.js]: Could return early; but, error carried through from previous

File: index.js (hunk 1 above)
Comment (roborourke) [reply index.js]: Cheers, fixed this one.

**Comment by roborourke** [comment]
As this is urgent-ish it'd be good to fix up the non-essential stuff like the logic carried over in a later PR.

**Pull request merged**